### PR TITLE
[Kernel] add fused_share_gate_sigmoid_mul jit kernel (cuda version)

### PR DESCRIPTION
**Pull request by Henson-Zh-Ali**
## Motivation && Modifications

Boost the performance of Qwen3.5 series (e.g., Qwen3.5-397B-A17B) by fusing the share_gate + sigmoid + multiply operations into one kernel. With this optimization, we got 1-2% uplift for throughput performance gain under production SLA limitation.

- before:
image

- after:
image

## Accuracy Tests

- Unit Tests:

```python
pytest python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py
```

image

- E2E Tests:

```python
python3 benchmark/mmmu/bench_sglang.py --port 9000
```

image

## Benchmark

Compared to the implementation (a pure PyTorch) currently used in SGL, this fused kernel reduces latency by approximately 100%–300% under various workloads (as tested on H20). Reproduction command:

```python
python3 python/sglang/jit_kernel/benchmark/bench_fused_share_gate_sigmoid_mul.py
```

Results:

image
image

### CODE REVIEW

File: python/sglang/jit_kernel/csrc/moe/fused_share_gate_sigmoid_mul.cuh [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+#pragma unroll
+    for (auto j = 0u; j < 4; ++j) {
+      const auto [x0, x1] = cast<fp32x2_t>(hidden_state_vec[j]);
+      const auto [y0, y1] = cast<fp32x2_t>(share_gate_weight_vec[j]);
+      gate_value += x0 * y0;
+      gate_value += x1 * y1;
+    }
+  }
+  gate_value = warp::reduce_sum(gate_value);
+  gate_value = __shfl_sync(0xffffffff, gate_value, 0);
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/moe/fused_share_gate_sigmoid_mul.cuh]: Why do we need this sync here? After `warp::reduce_sum`, each thread should hold the same value.

File: python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    import torch.nn.functional as F
+    share_expert_gate = torch.nn.Linear(share_gate_weight.shape[-1], 1, bias=False)
+    share_expert_gate.weight = torch.nn.Parameter(share_gate_weight)
+    x = share_expert_gate(hidden_state)
+    x = F.sigmoid(x)
+    x = x * share_expert_output
+    return x
+
+
+BS_LIST = [1, 7, 110, 1024, 4096, 65535]
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py]: Could we test more batch sizes? I would recommend scanning powers of 2 from 1 to 16384.

File: python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+HIDDEN_SIZE_LIST = [1024, 2048, 4096]
+DEVICE_LIST = ["cuda", ]
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py]: Give that currently all JIT kernels are written in CUDA, we may remove this?

File: python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    device: str,
+    dtype: torch.dtype,
+) -> None:
+    hidden_state = torch.randn(num_tokens, hidden_size, dtype=dtype, device=device)
+    share_gate_weight = torch.randn(1, hidden_size, dtype=dtype, device=device)
+    share_expert_output = torch.randn(num_tokens, hidden_size, dtype=dtype, device=device)
+    
+    ref_out = ref_fused_share_gate_sigmoid_mul(hidden_state, share_gate_weight, share_expert_output)
+    out = jit_fused_share_gate_sigmoid_mul(hidden_state, share_gate_weight, share_expert_output)
+    assert torch.allclose(ref_out, out, atol=5e-2, rtol=5e-3)
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py]: Why use `5e-2`? It looks too big to me. Typically for fp16/bf16, we use `1e-2`. You may refer to other JIT kernel tests as examples

Comment (Henson-Zh-Ali) [reply python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py]: Agree! And I confirm that this unit test can pass with using `1e-2`. Thanks for your suggestions.

File: python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py (hunk 3 above)
Comment (Henson-Zh-Ali) [reply python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py]: Perhaps the same implementation could also support other devices?

File: python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py (hunk 2 above)
Comment (Henson-Zh-Ali) [reply python/sglang/jit_kernel/tests/test_fused_share_gate_sigmoid_mul.py]: Agree~

File: python/sglang/jit_kernel/csrc/moe/fused_share_gate_sigmoid_mul.cuh (hunk 1 above)
Comment (Henson-Zh-Ali) [reply python/sglang/jit_kernel/csrc/moe/fused_share_gate_sigmoid_mul.cuh]: Agree~

**Comment by Henson-Zh-Ali** [comment]
@DarkSharpness hi, a polite ping~ could you please help me start the workflow if it looks good?

**Review by yuan-luo — APPROVED** [review]
I've reviewed this PR. This kernel's implementation is solid and the performance is pronounced. Thanks.

**Comment by hnyls2002** [comment]
Thanks @Henson-Zh-Ali! This fusion is now on main: `fused_gate_sigmoid_mul_add` in `python/sglang/kernels/ops/elementwise/elementwise.py:503` computes `final_hidden_states += sigmoid(hidden_states @ gate_weight) * shared_output` and is wired into the Qwen2-MoE shared-expert path at `python/sglang/srt/models/qwen2_moe.py:595`. Separately, `python/sglang/jit_kernel/` was relocated to `python/sglang/kernels/jit` + `kernels/ops`, so the added files no longer have a home. Please reopen if I've missed something.
